nautilus_core/risk/src/engine/mod.rs

```diff
@@ -22,140 +22,916 @@
 use std::{cell::RefCell, collections::HashMap, rc::Rc};
 
 use config::RiskEngineConfig;
-use nautilus_common::{cache::Cache, clock::Clock, msgbus::MessageBus, throttler::Throttler};
+use nautilus_common::{
+    cache::Cache,
+    clock::Clock,
+    logging::{CMD, EVT, RECV},
+    msgbus::MessageBus,
+    throttler::Throttler,
+};
+use nautilus_core::uuid::UUID4;
 use nautilus_execution::messages::{
     modify::ModifyOrder, submit::SubmitOrder, submit_list::SubmitOrderList, TradingCommand,
 };
 use nautilus_model::{
-    enums::TradingState,
-    events::order::OrderEventAny,
+    accounts::{any::AccountAny, base::Account},
+    enums::{InstrumentClass, OrderSide, OrderStatus, TradingState},
+    events::order::{OrderDenied, OrderEventAny, OrderModifyRejected},
     identifiers::InstrumentId,
     instruments::any::InstrumentAny,
     orders::{any::OrderAny, list::OrderList},
-    types::quantity::Quantity,
+    types::{currency::Currency, money::Money, price::Price, quantity::Quantity},
 };
-use rust_decimal::Decimal;
+use rust_decimal::{prelude::ToPrimitive, Decimal};
+use ustr::Ustr;
 
 pub mod config;
+// pub mod tests;
+
+type SubmitOrderFn = Box<dyn Fn(SubmitOrder)>;
+type ModifyOrderFn = Box<dyn Fn(ModifyOrder)>;
 
-pub struct RiskEngine<C>
-where
-    C: Clock,
-{
-    clock: C,
+pub struct RiskEngine {
+    clock: Rc<RefCell<dyn Clock>>,
     cache: Rc<RefCell<Cache>>,
     msgbus: Rc<RefCell<MessageBus>>,
-    order_submit_throttler: Throttler<SubmitOrder, Box<dyn Fn(SubmitOrder)>>,
-    order_modify_throttler: Throttler<ModifyOrder, Box<dyn Fn(ModifyOrder)>>,
+    // Counters
+    // command_count: u64,
+    // event_count: u64,
+    pub throttled_submit_order: Throttler<SubmitOrder, SubmitOrderFn, SubmitOrderFn>,
+    pub throttled_modify_order: Throttler<ModifyOrder, ModifyOrderFn, ModifyOrderFn>,
     max_notional_per_order: HashMap<InstrumentId, Decimal>,
+    trading_state: TradingState,
     config: RiskEngineConfig,
 }
 
-impl<C> RiskEngine<C>
-where
-    C: Clock,
-{
+impl RiskEngine {
+    pub fn new(
+        config: RiskEngineConfig,
+        // portfolio: PortfolioFacade TODO: fix after portfolio implementation
+        clock: Rc<RefCell<dyn Clock>>,
+        cache: Rc<RefCell<Cache>>,
+        msgbus: Rc<RefCell<MessageBus>>,
+    ) -> Self {
+        let msgbus1 = msgbus.clone();
+        let throttled_submit_order = Throttler::new(
+            config.max_order_submit.clone(),
+            clock.clone(),
+            "ORDER_SUBMIT_THROTTLER".to_string(),
+            Box::new(move |order: SubmitOrder| {
+                msgbus1
+                    .borrow_mut()
+                    .send(&Ustr::from("ExecEngine.execute"), &order);
+            }) as Box<dyn Fn(SubmitOrder)>,
```

Comment:
The trick here is that the closure does not need the whole RiskEngine i.e. `&self` state. It only needs a reference to the message bus since its own job is to send a message. By cloning and moving the message bus into the closure we can avoid the inner outer pattern.